### DISCUSSION TITLE
MediaPlayerElement in X.Skia.Gtk cause a build error

### DISCUSSION THREAD

**Discussion by Grzesik** [unoplatform/uno#12565] [General]
Hi, I added a MediaPlayerElement to a Xaml like this:
 <MediaPlayerElement x:Name="video" AutoPlay="True" AreTransportControlsEnabled="True" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" Background="AliceBlue"/>

It works on Windows, but on x.Skia.Gtk I have to add the package Uno.UI.MediaPlayer.Skia.Gtk to the x.Skia.Gtk project:

<PackageReference Include="Uno.UI.MediaPlayer.Skia.Gtk" />

When I compile the app (WSL Linux) I get an compiler error:

2>C:\Users\Marian.GRZESIK\.nuget\packages\uno.ui\4.9.20\buildTransitive\uno.winui.common.targets(23,3): error UNOB0001: Cannot build with both Uno.WinUI and Uno.UI nuget packages referenced. https://aka.platform.uno/UNOB0001
2>Done building project "UnoApp2.Skia.Gtk.csproj" -- FAILED.

The Version of the Uno:WinUi.xxx   are the same 4.9.20

**Comment by jeromelaban** [comment]
Thanks for the report. 

The documentation is not clear about what needs to be done, and you'll need to use the `Uno.WinUI.MediaPlayer.Skia.Gtk` as the app uses the WinUI APIs. We'll change the docs in https://github.com/unoplatform/uno/pull/12567.

**Comment by Grzesik** [comment]
Thanks Jerome, it solves the problem on Linux. 
I have to find now only the right path to the video on Linux to set it in the control. On Windows I could watch a video.

Thanks

**Reply by jeromelaban** [reply]
Thanks for the update. What type of URI are you using?

**Reply by Grzesik** [reply]
I've tried this code:


            string sLocalAppdata = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            var path = Path.Combine(sLocalAppdata, "video");
            path = Path.Combine(path, "video1.mp4");
            Uri pathUri = new Uri(path);
            video.Source = MediaSource.CreateFromUri(pathUri);


The path is on my local computer in WSL:

path = "/home/marian/.local/share/video/video1.mp4"

I put a video there, but it couldn't be displayed. I got this error:



```
Loaded '/mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/Uno.Core.Extensions.Logging.dll'. Cannot find or open the PDB file.
Exception thrown: 'System.DllNotFoundException' in LibVLCSharp.dll
Exception thrown: 'LibVLCSharp.Shared.VLCException' in LibVLCSharp.dll
fail: Uno.UI.Media.GtkMediaPlayer[0]
      Failed to initialize libVLC. See for more details: https://aka.platform.uno/mediaplayerelement
      LibVLCSharp.Shared.VLCException: LibVLC could not be created. Make sure that you have done the following:
      - Installed latest LibVLC from nuget for your target platform.
      Unable to load shared library 'libvlc' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
      /mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/runtimes/linux-x64/native/libvlc.so: cannot open shared object file: No such file or directory
      /usr/share/dotnet/shared/Microsoft.NETCore.App/7.0.5/libvlc.so: cannot open shared object file: No such file or directory
      /mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/libvlc.so: cannot open shared object file: No such file or directory
      /mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/runtimes/linux-x64/native/liblibvlc.so: cannot open shared object file: No such file or directory
      /usr/share/dotnet/shared/Microsoft.NETCore.App/7.0.5/liblibvlc.so: cannot open shared object file: No such file or directory
      /mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/liblibvlc.so: cannot open shared object file: No such file or directory
      /mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/runtimes/linux-x64/native/libvlc: cannot open shared object file: No such file or directory
      /usr/share/dotnet/shared/Microsoft.NETCore.App/7.0.5/libvlc: cannot open shared object file: No such file or directory
      /mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/libvlc: cannot open shared object file: No such file or directory
      /mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/runtimes/linux-x64/native/liblibvlc: cannot open shared object file: No such file or directory
      /usr/share/dotnet/shared/Microsoft.NETCore.App/7.0.5/liblibvlc: cannot open shared object file: No such file or directory
      /mnt/d/temp/UnoApp2/UnoApp2/UnoApp2.Skia.Gtk/bin/Debug/net7.0/liblibvlc: cannot open shared object file: No such file or directory
          at LibVLCSharp.Shared.Core.Native.LibVLCVersion()
         at LibVLCSharp.Shared.Core.EnsureVersionsMatch()
         at LibVLCSharp.Shared.Helpers.MarshalUtils.CreateWithOptions(String[] options, Func`3 create)
         at LibVLCSharp.Shared.Helpers.MarshalUtils.CreateWithOptions(String[] options, Func`3 create)
         at LibVLCSharp.Shared.LibVLC.<>c__DisplayClass9_0.<.ctor>b__0()
         at LibVLCSharp.Shared.Internal..ctor(Func`1 create, Action`1 release)
         at LibVLCSharp.Shared.LibVLC..ctor(Boolean enableDebugLogs, String[] options)
         at Uno.UI.Media.GtkMediaPlayer.<InitializeInner>b__85_0()
```


I called the following command in my Wsl/Ubuntu:

```
sudo apt-get install libvlc-dev libx11-dev vlc libgtk2.0-0 libx11dev

Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package libx11dev
```

Something is here not really correct

**Reply by jeromelaban** [reply]
Thanks for the update. It seems that libx11dev is not available in your context, or maybe you're missing a `sudo apt update` to run first.

Try running:
```
sudo apt update
sudo apt install libvlc-dev libx11-dev vlc libgtk2.0-0
sudo apt install libx11dev
```

**Reply by Grzesik** [reply]
I have still the problem. I've found in this link https://itsfoss.com/unable-to-locate-package-error-ubuntu/    that the package libx11dev doesn't exist in my Ubuntu:

marian@P-GRZESIK-AMD:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal

Is the package libx11dev really correct?

**Reply by jeromelaban** [reply]
Thanks. It may not be needed for your distribution. Does the app work the with the new installed packages?

**Comment by Grzesik** [comment]
Unfortunately not. I get two exceptions:

1. Exception with the path.

Exception thrown: 'System.InvalidOperationException' in System.Private.Uri.dll
Loaded '/usr/share/dotnet/shared/Microsoft.NETCore.App/7.0.5/System.IO.MemoryMappedFiles.dll'. Cannot find or open the PDB file.
fail: Microsoft.UI.Xaml.Data.BindingExpression[0]
      Failed to apply binding to property [DependencyPropertyDetails(MediaPlayer)] on [Microsoft.UI.Xaml.Controls.MediaPlayerPresenter] (This operation is not supported for a relative URI.)
      System.InvalidOperationException: This operation is not supported for a relative URI.
         at System.Uri.get_AbsoluteUri()


2. (UnoApp2.Skia.Gtk:3011): Gdk-WARNING **: 15:39:40.224: ../../../../../gdk/x11/gdkwindow-x11.c:5633 drawable is not a native X11 window
fail: Uno.UI.Media.VideoView[0]
      Unable to get the X11 Window ID. This may be caused when running in a Wayland environment, such as WSLg. https://aka.platform.uno/mediaplayerelement
Exception thrown: 'System.InvalidOperationException' in System.Private.Uri.dll
Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll
fail: Uno.UI.Dispatching.CoreDispatcher[0]
      Dispatcher unhandled exception
      System.InvalidOperationException: This operation is not supported for a relative URI.


I'll try it on Raspberry Pi. It is for me the target platform. I wanted only test it with WSL. But I think, I can test it also with Windows and than copy the programm to Raspberry Pi

**Reply by jeromelaban** [reply]
Thanks. So there are two issues here, the first is that the URL you're providing is not supported somehow. What is it? For the second issue, WSL 2 is not supported at this time (it uses Wayland that libVLC does not support at this time).

**Comment by Grzesik** [comment]
Hi Jerome (FYI),
I have tried to run a simple app with the  MediaPlayerElement on the Raspberry Pi. My first simple sample works fine. I had only problems, how to reference a file from a file system.

Setps:
1. 64 Bit Os with desktop
2. Run this commands:
    sudo apt update
    sudo apt install libvlc-dev libx11-dev vlc libgtk2.0-0
3. Set the display
    export DISPLAY=:0.0
4. Run my app (arm64 self-contained)
   ./RaspAdvertising.Skia.Gtk

To reference a file from a file system, it works like this:

video.Source = MediaSource.CreateFromUri(new Uri("file:///home/pi/Videos/video1.mp4"));

The next steps are, to display the app in fullscreen mode. forever loop and a kiosk mode.

**Reply by jeromelaban** [reply]
Thanks for the report! Support for local files is being worked on here: https://github.com/unoplatform/uno/pull/12606
